proposals/0174-sbpf-arithmetics-improvements.md

```diff
@@ -0,0 +1,154 @@
+category: Standard
+type: Core
+status: Draft
+created: 2024-09-06
+feature: TBD
+extends: SIMD-0161
+---
+
+## Summary
+
+This proposal introduces wide multiplication, signed division and explicit sign
+extension to SBPF.
+
+## Motivation
+
+All major hardware ISAs support 64 x 64 = 128 bit multiplication; BPF does not.
+This hurts performance of big integer arithmetics. Similarly, signed division
+and signed remainder / modulo instructions must be emulated in software as
+well.
+
+Another issue related to arithmetics is that some 32 bit instructions perform
+sign extension of their results (output, **not** input, there is a difference)
+implicitly. This is first of all useless because source languages do not read
+the 32 MSBs of a 32 bit result stored in a 64 bit register. And second, it
+requires interpreters and compilers to perform extra work to add sign extension
+at the end of these instructions. Instead we should go the same route as the
+underlying hardware ISAs and require sign extension to be made explicit in a
+dedicated instruction.
+
+Furthermore, the instruction `sub dst, imm` is redundant as it can also be
+encoded as `add dst, -imm`. If we were to swap the operands meaning of minuend
+and subtrahend, then the instruction would become useful and would even render
+`neg dst` redundant. Negation would then be encoded as `reg = 0 - reg`. This
+would also make it clear how sign extension rules work for negation.
+
+## Alternatives Considered
+
+None.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+The following must go into effect if and only if a program indicates the
+SBPF-version (TBD) or higher in its program header (see SIMD-0161).
+
+### Changes to the Bytecode Verifier
+
+A program containing one of the following instructions must throw
+`VerifierError::UnknownOpCode` during verification:
+
+- the `MUL` instruction (opcodes `0x24`, `0x2C`, `0x27` and `0x2F`)
+- the `DIV` instruction (opcodes `0x34`, `0x3C`, `0x37` and `0x3F`)
+- the `MOD` instruction (opcodes `0x94`, `0x9C`, `0x97` and `0x9F`)
+- the `NEG` instruction (opcodes `0x84` and `0x87`)
+
+A program containing one of the following instructions must **not** throw
+`VerifierError::UnknownOpCode` during verification anymore (opcodes are for
+immediate and register variant each, in that order):
+
+- the `UHMUL64` instruction (opcode `0x36` and `0x3E`)
+- the `UDIV32` instruction (opcode `0x46` and `0x4E`)
+- the `UDIV64` instruction (opcode `0x56` and `0x5E`)
+- the `UREM32` instruction (opcode `0x66` and `0x6E`)
+- the `UREM64` instruction (opcode `0x76` and `0x7E`)
+- the `LMUL32` instruction (opcode `0x86` and `0x8E`)
+- the `LMUL64` instruction (opcode `0x96` and `0x9E`)
+- the `SHMUL64` instruction (opcode `0xB6` and `0xBE`)
+- the `SDIV32` instruction (opcode `0xC6` and `0xCE`)
+- the `SDIV64` instruction (opcode `0xD6` and `0xDE`)
+- the `SREM32` instruction (opcode `0xE6` and `0xEE`)
+- the `SREM64` instruction (opcode `0xF6` and `0xFE`)
```

Comment:
I think these instruction opcodes is a bit of a departure from the pattern we had until now. In BPF, opcode for 32-bit and 64-bit instructions is in the least significant bits: `mul64 dest, reg` is `0x2F` and `mul32 dest, reg` is `0x2C`.

For these new instructions, we change the most significant bits. Maintaining the same pattern simplifies code generation.

Reply:
Yes, but that bit is part of the instruction class. Meaning the 32 bit ALU instruction class and the 64 bit ALU instruction class are two different instruction classes. And this SIMD introduces a new instruction class for PQR because they don't fit into the ALU classes.